docs/ContainerRuntimeConfigDesign.md

```diff
@@ -0,0 +1,95 @@
+
+The CRI-O runtime is used by default and is the recommended solution.
+
+# Motivation
+
+The MCO contains the logic to write the crio.conf, storage.conf, and registries.conf configuration files to Ignition. When Ignition starts on a machine it writes these three files to configure the container runtime. When these resources change within the MachineConfig, the new configs are rewritten and the MachineConfigDaemon is instructed to reboot the machine to pick up the new configs.
+
+## Goals
+
+The following goals should not require explicit knowledge of a customer to know the right command-line or configuration options to use within the ContainerRuntime.
+
+1. setting pids limit
+
+2. setting log level
+
+3. configuring list of search registries
+
+4. configuring list of insecure registries
+
+5. configuring list of blocked registries
+
+6. configuring max size of storage driver
+
+## Non-Goals
+
+# Proposal
+
+Extend the Machine Config Operator to include a ContainerRuntimeConfig CRD and ContainerRuntimeConfigController. By using a ContainerRuntimeConfig CRD there is an implicit whitelist of user controlled options for the container runtime. Upon deleting the ContainerRuntimeConfig instance the default config is restored*.
+
+## CRD
+
+```
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+ name: containerruntimeconfigs.machineconfiguration.openshift.io
+spec:
+ group: machineconfiguration.openshift.io
+ versions:
+   - name: v1
+     served: true
+     storage: true
+ scope: Cluster
+ names:
+   plural: containerruntimeconfigs
+   singular: containerruntimeconfig
+   kind: ContainerRuntimeConfig
+```
+
+## SPEC
+
+```
+MachineConfigPoolSelector *metav1.LabelSelector
+Runtime:
+ Name  string
+ Endpoint string
+ContainerRuntimeConfig
+ [ContainerRuntimeConfigurationSpec](Will have to add this in this repo)
+```
+
+Example:
+
+```
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+ name: set-pids-limit
+spec:
+ machineConfigPoolSelector:
+   matchLabels:
+     custom-crio: high-pid-limit
+ containerRuntimeConfig:
+   pidsLimit: 2048
+```
+
+## Implementation Details
+
+The ContainerRuntimeConfigController would perform the following steps:
+
+1. Validate the user defined ContainerRuntimeConfig
+
+2. Render the current MachineConfigs (storage.files.contents[crio.conf, storage.conf, registries.conf]) into the ContainerRuntimeConfiguration structure
+
+3. Load the ContainerRuntimeConfig from the passed in Spec.ContainerRuntimeConfig
+
+4. Serialize the ContainerRuntimeConfig to yaml
+
+5. Create or Update the ignition /etc/containers/storage.conf, /etc/containers/registries.conf, and /etc/crio/crio.conf files within a 99-[role]-kubelet-managed MachineConfig
```

Comment:
Hmm, why 99 here? This just has to be > 00 to override the base templates, right? The connotation with "99" is that no MC should override these files but we want to have that flexibility, right? (Of course, one could always have a "999", but that doesn't look as nice :)).

Reply:
Right, it just has to be > 01 to override the base templates. Saw 99 being used, so just followed that. I can change it to 02.

Reply:
I chose 99- for a managed MachineConfig. The default crio.conf should probably be at 01- to match the kubelet conf default machineconfig.